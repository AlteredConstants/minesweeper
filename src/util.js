// @flow
import { flow, fill, map, shuffle, filter, sumBy } from 'lodash/fp';

const mapWithIndex = map.convert({ cap: false });

const distributeMines = count => flow(fill(0, count, true), shuffle);

const createTiles = width =>
  mapWithIndex((isMine, index: number): Mine$Tile => ({
    index,
    row: Math.floor(index / width),
    column: index % width,
    isMine,
    adjacentMineCount: 0,
    isCleared: false,
    isFlagged: false,
  }));

const getAdjacentCoordinates = ({ row, column }) => {
  const top = row - 1;
  const bottom = row + 1;
  const left = column - 1;
  const right = column + 1;
  return [
    [top, left],
    [top, column],
    [top, right],
    [row, left],
    [row, right],
    [bottom, left],
    [bottom, column],
    [bottom, right],
  ];
};

const getAdjacentTiles = ({ width, height, tiles }) =>
  flow(
    getAdjacentCoordinates,
    filter(
      ([row, column]) =>
        row >= 0 && row < height && column >= 0 && column < width,
    ),
    map(([row, column]) => tiles[row * width + column]),
  );

const updateAdjacentMineCount = (width, height) =>
  mapWithIndex((tile, index: number, tiles) => ({
    ...tile,
    adjacentMineCount: flow(
      getAdjacentTiles({ width, height, tiles }),
      sumBy(adjacentTile => (adjacentTile.isMine ? 1 : 0)),
    )(tile),
  }));

export function createField(options: {
  width: number,
  height: number,
  mineCount: number,
}): Mine$Field {
  const { width, height, mineCount } = options;
  const size = width * height;
  const tiles = flow(
    fill(0, size, false),
    distributeMines(mineCount),
    createTiles(width),
    updateAdjacentMineCount(width, height),
  )(Array(size));
  return {
    width,
    height,
    mineCount,
    tiles,
    isExploded: false,
  };
}

function* connectedSafeTilesGenerator(
  field: Mine$Field,
  tile: Mine$Tile,
  seenTiles: Map<number, boolean> = new Map(),
): Iterable<Mine$Tile> {
  if (seenTiles.get(tile.index)) {
    return;
  }
  seenTiles.set(tile.index, true);
  yield tile;
  if (tile.adjacentMineCount !== 0) {
    return;
  }
  for (let adjacentTile of getAdjacentTiles(field)(tile)) {
    yield* connectedSafeTilesGenerator(field, adjacentTile, seenTiles);
  }
}

export function getConnectedSafeTiles(field: Mine$Field, tile: Mine$Tile) {
  return [...connectedSafeTilesGenerator(field, tile)];
}

function matches(object, newProps) {
  if (object === newProps) {
    return false;
  }
  if (object == null || newProps == null) {
    return true;
  }
  return Object.entries(newProps).reduce(
    (result, [key, value]) => result || object[key] === value,
    false,
  );
}

export function updateInObject<T: { [string]: * }, U: { [string]: T }>(
  object: U,
  key: string,
  newProps: $Shape<T>, // TODO: This doesn't work: https://github.com/facebook/flow/issues/3713
) {
  return matches(object[key], newProps)
    ? object
    : { ...object, [key]: { ...object[key], ...newProps } };
}

export function updateInArray<T: { [string]: * }, U: $ReadOnlyArray<T>>(
  array: U,
  indexes: number | Array<number>,
  newProps: $Shape<T>, // TODO: This doesn't work: https://github.com/facebook/flow/issues/3713
) {
  const updateIndexes = Array.isArray(indexes) ? indexes : [indexes];
  let isUpdated = false;
  const newArray = array.map((item, index) => {
    if (!updateIndexes.includes(index) || matches(array[index], newProps)) {
      return item;
    }
    isUpdated = true;
    return { ...item, ...newProps };
  });
  return isUpdated ? newArray : array;
}
