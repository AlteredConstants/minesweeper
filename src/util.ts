import { flow, fill, map, shuffle, filter, sumBy } from 'lodash/fp';
import { Field, Tile } from './index';

const mapFull = map.convert({ cap: false });

const distributeMines = (count: number) => flow(fill(0, count, true), shuffle);

const createTiles = (width: number) =>
  mapFull((isMine: boolean, index): Tile => ({
    index,
    row: Math.floor(index / width),
    column: index % width,
    isMine: isMine,
    isCleared: false,
    adjacentMineCount: 0,
    isFlagged: false,
  }));

const getAdjacentCoordinates = ({ row, column }: Tile) => {
  const top = row - 1;
  const bottom = row + 1;
  const left = column - 1;
  const right = column + 1;
  return [
    [top, left],
    [top, column],
    [top, right],
    [row, left],
    [row, right],
    [bottom, left],
    [bottom, column],
    [bottom, right],
  ];
};

const getAdjacentTiles = ({
  width,
  height,
  tiles,
}: { width: number; height: number; tiles: ReadonlyArray<Tile> }) =>
  flow(
    getAdjacentCoordinates,
    filter(
      ([row, column]) =>
        row >= 0 && row < height && column >= 0 && column < width,
    ),
    map(([row, column]) => tiles[row * width + column]),
  );

const updateAdjacentMineCount = (width: number, height: number) =>
  mapFull((tile: Tile, _, tiles: Tile[]) => ({
    ...tile,
    adjacentMineCount: flow(
      getAdjacentTiles({ width, height, tiles }),
      sumBy((adjacentTile: Tile) => (adjacentTile.isMine ? 1 : 0)),
    )(tile),
  }));

export function createField(
  options: { width: number; height: number; mineCount: number },
): Field {
  const { width, height, mineCount } = options;
  const size = width * height;
  const tiles = flow(
    fill(0, size, false),
    distributeMines(mineCount),
    createTiles(width),
    updateAdjacentMineCount(width, height),
  )(Array<boolean>(size));
  return {
    width,
    height,
    mineCount,
    tiles,
    isExploded: false,
  };
}

function* connectedSafeTilesGenerator(
  field: Field,
  tile: Tile,
  seenTiles = new Map<number, boolean>(),
): Iterable<Tile> {
  if (seenTiles.get(tile.index)) {
    return undefined;
  }
  seenTiles.set(tile.index, true);
  yield tile;
  if (tile.adjacentMineCount !== 0) {
    return undefined;
  }
  for (let adjacentTile of getAdjacentTiles(field)(tile)) {
    yield* connectedSafeTilesGenerator(field, adjacentTile, seenTiles);
  }
}

export function getConnectedSafeTiles(field: Field, tile: Tile) {
  return [...connectedSafeTilesGenerator(field, tile)];
}
