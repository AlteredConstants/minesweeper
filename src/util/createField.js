import { flow, map, sumBy } from "lodash/fp";
import distributeMines from "./distributeMines";
import { createAdjacentTilesGetter } from "./getAdjacentTiles";

const uncappedMap = map.convert({ cap: false });

const createTilesCreator = width =>
  uncappedMap((isMine, index) => ({
    index,
    row: Math.floor(index / width),
    column: index % width,
    isMine,
    adjacentMineCount: 0,
    isCleared: false,
    isFlagged: false,
  }));

const createAdjacentMineCountUpdater = (width, height) =>
  uncappedMap((tile, _, tiles) => ({
    ...tile,
    adjacentMineCount: flow(
      createAdjacentTilesGetter({ width, height, tiles }),
      sumBy(adjacentTile => (adjacentTile.isMine ? 1 : 0)),
    )(tile),
  }));

export default function createField(options) {
  const { width, height, mineCount } = options;
  const createTiles = createTilesCreator(width);
  const updateAdjacentMineCount = createAdjacentMineCountUpdater(width, height);
  const mines = distributeMines(width * height, mineCount);
  const tiles = updateAdjacentMineCount(createTiles(mines));
  return {
    width,
    height,
    mineCount,
    tiles,
    isExploded: false,
  };
}
