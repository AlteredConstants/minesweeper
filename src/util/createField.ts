import { flow, map, sumBy } from "lodash/fp";
import { Field, FieldState, Tile } from "../interface";
import distributeMines from "./distributeMines";
import { createAdjacentTilesGetter } from "./getAdjacentTiles";

const uncappedMap = map.convert({ cap: false });

const createTilesCreator = (width: number) =>
  uncappedMap((isMine: boolean, index): Tile => ({
    index,
    row: Math.floor(index / width),
    column: index % width,
    isMine,
    adjacentMineCount: 0,
    isCleared: false,
    isFlagged: false,
  }));

export const updateAdjacentMineCount = uncappedMap(
  (tile: Tile, index, tiles: Tile[]): Tile => ({
    ...tile,
    adjacentMineCount: flow(
      createAdjacentTilesGetter(tiles),
      sumBy(adjacentTile => (adjacentTile.isMine ? 1 : 0)),
    )(tile),
  }),
);

export default function createField(options: {
  width: number;
  height: number;
  mineCount: number;
}): Field {
  const { width, height, mineCount } = options;
  const createTiles = createTilesCreator(width);
  const mines = distributeMines(width * height, mineCount);
  const tiles = updateAdjacentMineCount(createTiles(mines));
  return {
    width,
    height,
    mineCount,
    tiles,
    state: FieldState.Active,
  };
}

export type FieldCreator = typeof createField;
