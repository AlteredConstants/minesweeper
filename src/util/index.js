export { createField } from "./createField";
export { default as getConnectedSafeTiles } from "./getConnectedSafeTiles";
export { updateInObject, updateInArray } from "./updateIn";
