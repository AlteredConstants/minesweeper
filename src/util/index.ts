export { default as createField } from "./createField";
export { default as getConnectedSafeTiles } from "./getConnectedSafeTiles";
export { updateInObject, updateInArray } from "./updateIn";
export { default as getAdjacentTiles } from "./getAdjacentTiles";

export const getCoordinates = (index: number, width: number) => ({
  row: Math.floor(index / width),
  column: index % width,
});
