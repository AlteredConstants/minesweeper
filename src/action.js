// @flow
import { createField, getConnectedSafeTiles } from './util';
import type { State } from './reducer';

const DefaultField = { width: 30, height: 16, mineCount: 99 };

type GetState = () => State;
type Dispatch<A> = (action: A) => void;
type ThunkAction<A> = (dispatch: Dispatch<A>, getState: GetState) => void;

type StartNewFieldAction = {
  type: 'START_NEW_FIELD',
  field: Field,
};
export function startNewField(
  options: {
    width: number,
    height: number,
    mineCount: number,
  } = DefaultField,
): StartNewFieldAction {
  return {
    type: 'START_NEW_FIELD',
    field: createField(options),
  };
}

type ToggleFlagTileAction = {
  type: 'TOGGLE_FLAG_TILE',
  tile: Tile,
};
export function toggleFlagTile(tile: Tile): ToggleFlagTileAction {
  return {
    type: 'TOGGLE_FLAG_TILE',
    tile,
  };
}

type ClearTileAction = {
  type: 'CLEAR_TILE',
  tile: Tile,
};
export function clearTile(tile: Tile): ClearTileAction {
  return {
    type: 'CLEAR_TILE',
    tile,
  };
}

type ClearConnectSafeTilesAction = {
  type: 'CLEAR_CONNECTED_SAFE_TILES',
  tiles: Array<Tile>,
};
export function clearConnectedSafeTiles(
  field: Field,
  tile: Tile,
): ClearConnectSafeTilesAction {
  return {
    type: 'CLEAR_CONNECTED_SAFE_TILES',
    tiles: getConnectedSafeTiles(field, tile),
  };
}

type TripMineAction = {
  type: 'TRIP_MINE',
};
export function tripMine(): TripMineAction {
  return {
    type: 'TRIP_MINE',
  };
}

type ClearAction = ThunkAction<
  ClearTileAction | ClearConnectSafeTilesAction | TripMineAction
>;
export function clear(tile: Tile): ClearAction {
  return (dispatch, getState) => {
    dispatch(clearTile(tile));
    if (tile.isCleared) {
      return;
    }
    if (tile.isMine) {
      dispatch(tripMine());
    } else if (tile.adjacentMineCount === 0) {
      const { field } = getState();
      if (field) {
        dispatch(clearConnectedSafeTiles(field, tile));
      }
    }
  };
}

export type Action =
  | ClearConnectSafeTilesAction
  | ClearTileAction
  | StartNewFieldAction
  | ToggleFlagTileAction
  | TripMineAction;
