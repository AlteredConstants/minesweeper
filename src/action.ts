import { Action } from "redux";
import { Field, Tile } from "./interface";
import { createField } from "./util";

const DefaultField = { width: 30, height: 16, mineCount: 99 };

export type Action =
  | ClearAdjacentTilesAction
  | ClearTileAction
  | StartNewFieldAction
  | ToggleFlagTileAction;

interface StartNewFieldAction extends Action {
  type: "START_NEW_FIELD";
  field: Field;
}
export function startNewField(
  options: {
    width: number;
    height: number;
    mineCount: number;
  } = DefaultField,
): StartNewFieldAction {
  return {
    type: "START_NEW_FIELD",
    field: createField(options),
  };
}

interface ToggleFlagTileAction extends Action {
  type: "TOGGLE_FLAG_TILE";
  tile: Tile;
}
export function toggleFlagTile(tile: Tile): ToggleFlagTileAction {
  return {
    type: "TOGGLE_FLAG_TILE",
    tile,
  };
}

interface ClearTileAction extends Action {
  type: "CLEAR_TILE";
  tile: Tile;
}
export function clearTile(tile: Tile): ClearTileAction {
  return {
    type: "CLEAR_TILE",
    tile,
  };
}

interface ClearAdjacentTilesAction extends Action {
  type: "CLEAR_ADJACENT_TILES";
  tile: Tile;
}
export function clearAdjacentTiles(tile: Tile): ClearAdjacentTilesAction {
  return {
    type: "CLEAR_ADJACENT_TILES",
    tile,
  };
}
