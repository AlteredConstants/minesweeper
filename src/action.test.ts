import * as creator from "./action";
import { createField } from "./util";
import {
  mockConnectedSafeTilesForZeroTile,
  mockField,
  mockMineTile,
  mockNumberTile,
  mockSurroundedNumberTile,
  mockZeroTile,
} from "./util/__mocks__/createField";

jest.mock("./util/getConnectedSafeTiles");

describe("startNewField", () => {
  it("should call createField", () => {
    creator.startNewField("<Mock options>" as any, "<Mock index>" as any);
    expect(createField).toHaveBeenCalledWith("<Mock options>", "<Mock index>");
  });

  it("should create a START_NEW_FIELD action", () => {
    const action = creator.startNewField("<Mock options>" as any, 0);
    expect(action).toMatchSnapshot();
  });
});

describe("toggleFlagTile", () => {
  it("should create a TOGGLE_FLAG_TILE action", () => {
    const action = creator.toggleFlagTile(mockZeroTile);
    expect(action).toMatchSnapshot();
  });
});

describe("clearTile", () => {
  it("should create a CLEAR_TILE action", () => {
    const action = creator.clearTile(mockZeroTile);
    expect(action).toMatchSnapshot();
  });
});

describe("clearAdjacentTile", () => {
  it("should create a CLEAR_ADJACENT_TILE action", () => {
    const action = creator.clearTile(mockNumberTile);
    expect(action).toMatchSnapshot();
  });
});

// This action creator no longer exists.
// TODO: probably migrate these to the relevant reducer tests.
describe.skip("clear", () => {
  it("should not dispatch any actions if the tile is already cleared", () => {
    const action = creator.clear({ ...mockZeroTile, isCleared: true });
    const dispatch = jest.fn();
    const getState = jest.fn(() => ({ game: { field: mockField } }));
    action(dispatch, getState);
    expect(dispatch).not.toHaveBeenCalled();
  });

  it("should dispatch a CLEAR_TILE action", () => {
    const action = creator.clear(mockZeroTile);
    const dispatch = jest.fn();
    const getState = jest.fn(() => ({ game: { field: mockField } }));
    action(dispatch, getState);
    expect(dispatch).toHaveBeenCalledWith({
      type: "CLEAR_TILE",
      tile: mockZeroTile,
    });
  });

  it("should dispatch a TRIP_MINE action for a mine tile", () => {
    const action = creator.clear(mockMineTile);
    const dispatch = jest.fn();
    const getState = jest.fn(() => ({ game: {} }));
    action(dispatch, getState);
    expect(dispatch).toHaveBeenCalledWith({ type: "TRIP_MINE" });
  });

  it("should dispatch a CLEAR_CONNECTED_SAFE_TILES action for a tile with no adjacent mines", () => {
    const action = creator.clear(mockZeroTile);
    const dispatch = jest.fn();
    const getState = jest.fn(() => ({ game: { field: mockField } }));
    action(dispatch, getState);
    expect(dispatch).toHaveBeenCalledWith({
      type: "CLEAR_CONNECTED_SAFE_TILES",
      originTile: mockZeroTile,
      tiles: mockConnectedSafeTilesForZeroTile,
    });
  });

  it("should not dispatch a CLEAR_CONNECTED_SAFE_TILES action for a tile with adjacent mines", () => {
    const action = creator.clear(mockSurroundedNumberTile);
    const dispatch = jest.fn();
    const getState = jest.fn(() => ({ game: { field: mockField } }));
    action(dispatch, getState);
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({
      type: "CLEAR_TILE",
      tile: mockSurroundedNumberTile,
    });
  });
});
