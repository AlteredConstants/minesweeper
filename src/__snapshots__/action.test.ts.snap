// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clearAdjacentTile should create a CLEAR_ADJACENT_TILE action 1`] = `
Object {
  "tile": Object {
    "adjacentMineCount": 1,
    "column": 1,
    "index": 5,
    "isCleared": false,
    "isFlagged": false,
    "isMine": false,
    "row": 1,
  },
  "type": "CLEAR_TILE",
}
`;

exports[`clearTile should create a CLEAR_TILE action 1`] = `
Object {
  "tile": Object {
    "adjacentMineCount": 0,
    "column": 0,
    "index": 4,
    "isCleared": false,
    "isFlagged": false,
    "isMine": false,
    "row": 1,
  },
  "type": "CLEAR_TILE",
}
`;

exports[`startNewField should create a START_NEW_FIELD action 1`] = `
Object {
  "field": Object {
    "height": 4,
    "mineCount": 4,
    "state": "active",
    "tiles": Array [
      Object {
        "adjacentMineCount": 0,
        "column": 0,
        "index": 0,
        "isCleared": false,
        "isFlagged": false,
        "isMine": false,
        "row": 0,
      },
      Object {
        "adjacentMineCount": 0,
        "column": 1,
        "index": 1,
        "isCleared": false,
        "isFlagged": false,
        "isMine": false,
        "row": 0,
      },
      Object {
        "adjacentMineCount": 1,
        "column": 2,
        "index": 2,
        "isCleared": false,
        "isFlagged": false,
        "isMine": false,
        "row": 0,
      },
      Object {
        "adjacentMineCount": 1,
        "column": 3,
        "index": 3,
        "isCleared": false,
        "isFlagged": false,
        "isMine": false,
        "row": 0,
      },
      Object {
        "adjacentMineCount": 0,
        "column": 0,
        "index": 4,
        "isCleared": false,
        "isFlagged": false,
        "isMine": false,
        "row": 1,
      },
      Object {
        "adjacentMineCount": 1,
        "column": 1,
        "index": 5,
        "isCleared": false,
        "isFlagged": false,
        "isMine": false,
        "row": 1,
      },
      Object {
        "adjacentMineCount": 2,
        "column": 2,
        "index": 6,
        "isCleared": false,
        "isFlagged": false,
        "isMine": false,
        "row": 1,
      },
      Object {
        "adjacentMineCount": 1,
        "column": 3,
        "index": 7,
        "isCleared": false,
        "isFlagged": false,
        "isMine": true,
        "row": 1,
      },
      Object {
        "adjacentMineCount": 1,
        "column": 0,
        "index": 8,
        "isCleared": false,
        "isFlagged": false,
        "isMine": false,
        "row": 2,
      },
      Object {
        "adjacentMineCount": 2,
        "column": 1,
        "index": 9,
        "isCleared": false,
        "isFlagged": false,
        "isMine": false,
        "row": 2,
      },
      Object {
        "adjacentMineCount": 2,
        "column": 2,
        "index": 10,
        "isCleared": false,
        "isFlagged": false,
        "isMine": true,
        "row": 2,
      },
      Object {
        "adjacentMineCount": 3,
        "column": 3,
        "index": 11,
        "isCleared": false,
        "isFlagged": false,
        "isMine": false,
        "row": 2,
      },
      Object {
        "adjacentMineCount": 0,
        "column": 0,
        "index": 12,
        "isCleared": false,
        "isFlagged": false,
        "isMine": true,
        "row": 3,
      },
      Object {
        "adjacentMineCount": 2,
        "column": 1,
        "index": 13,
        "isCleared": false,
        "isFlagged": false,
        "isMine": false,
        "row": 3,
      },
      Object {
        "adjacentMineCount": 2,
        "column": 2,
        "index": 14,
        "isCleared": false,
        "isFlagged": false,
        "isMine": false,
        "row": 3,
      },
      Object {
        "adjacentMineCount": 1,
        "column": 3,
        "index": 15,
        "isCleared": false,
        "isFlagged": false,
        "isMine": true,
        "row": 3,
      },
    ],
    "width": 4,
  },
  "type": "START_NEW_FIELD",
}
`;

exports[`toggleFlagTile should create a TOGGLE_FLAG_TILE action 1`] = `
Object {
  "tile": Object {
    "adjacentMineCount": 0,
    "column": 0,
    "index": 4,
    "isCleared": false,
    "isFlagged": false,
    "isMine": false,
    "row": 1,
  },
  "type": "TOGGLE_FLAG_TILE",
}
`;
